new: Port Lithium's "fast level loading screen" patch

Leo40Git opened the pull request:
This is the same patch as seen in CaffeineMC/lithium-fabric@ac2219c, with a few changes:
* Updated for 1.16 (IE there's a `MatrixStack` now)
* Now using Sodium's `VertexSink` system!
* Added a new `BasicScreenQuad` vertex type, for colored quads
  * TODO: Use this new vertex type elsewhere (`ItemRenderer`, `DrawableHelper.fill`, `AbstractWidget`...)
  * TODO: Create textured variant and use *that* type elsewhere (`DrawableHelper`, some `Widget`s)

Leo40Git committed:
new: Port Lithium's "fast level loading screen" patch

This is the same patch as seen in https://github.com/CaffeineMC/lithium-fabric/commit/ac2219c7f460b3b4ede4cb75590250a5d4562d05, just updated for 1.16 (IE there's a MatrixStack now)

Leo40Git committed:
Move mixin into its own subpackage

Leo40Git committed:
Buffer is now expanded to expected size before operations begin

maityyy commented:
Color4 should be named Color4i

Leo40Git committed:
Implement new "screen quad" vertex type (name pending?)

This allows the level loading screen to use quicker intrinsics
Also removed Color4, since these intrinsics allow a packed ABGR color to be used directly

Leo40Git commented:
> `Color4` should be renamed `Color4i` to follow coding standards.

Well I kinda deleted it, so issue resolved!

Leo40Git committed:
Ensure enough capacity for all *vertices*, not quads

Leo40Git committed:
Correct colors

RGB -> ABGR, not BGR -> ABGR

Leo40Git committed:
Use ColorARGB.toABGR instead

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/mixin/features/gui/fast_loading_screen/MixinLevelLoadingScreen.java:
```diff
@@ -0,0 +1,133 @@
+            int mapRenderCenterSize = centerSize * tileSize - mapPadding;
+            int radius = mapRenderCenterSize / 2 + 1;
+
+            addRect(matrix, sink, mapX - radius, mapY - radius, mapX - radius + 1, mapY + radius, DEFAULT_STATUS_COLOR);
```
It would be better to create a `sink.ensureCapacity(4 * 4)` call in this if-body so that the logic above isn't needed. Then, later, another `sink.ensureCapacity(size * size * 4)` call can be made before the for-loops at line 98.

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/model/vertex/VanillaVertexTypes.java:
```diff
@@ -8,11 +8,14 @@
     public static final VanillaVertexType<LineVertexSink> LINES = new LineVertexType();
     public static final VanillaVertexType<GlyphVertexSink> GLYPHS = new GlyphVertexType();
     public static final VanillaVertexType<ParticleVertexSink> PARTICLES = new ParticleVertexType();
+    public static final VanillaVertexType<ScreenQuadVertexSink> SCREEN_QUADS = new ScreenQuadVertexType();
```
Many different "quad" formats are used for UI rendering, some which include light and textures. A better name for the format would be `UntexturedScreenQuad` or `BasicScreenQuad`... etc.

Leo40Git replied:
This can be `BasicScreenQuad`, then a version with textures can be `TexturedScreenQuad`.
If you want `TexturedScreenQuad` with lighting, `GlyphVertexSink` (since I'm pretty sure that's the only UI-related type with a light component).

Leo40Git committed:
JellySquid changes

- Renamed ScreenQuad to BasicScreenQuad
- ensureCapacity is now called as needed, rather than at the start

jellysquid3 reviewed (approved):
LGTM.
